File: AdventOfCode2023/Day20.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AdventOfCode2023
{
    internal class Day20 : Day
    {
        class Node
        {
            class PulseData
            {
                public string source;
                public string dest;
                public bool high;
            }

            public static void SendPulse(bool high, string source, string dest)
            {
                PulseData data = new PulseData();
                data.source = source;
                data.dest = dest;
                data.high = high;
                PulseQueue.Enqueue(data);

                if (high)
                    highPulses++;
                else
                    lowPulses++;
            }

            public static bool ProcessQueue()
            {
                if (PulseQueue.Count == 0)
                    return false;

                PulseData data = PulseQueue.Dequeue();

                nodes[data.dest].Pulse(data.high, data.source);

                return true;
            }

            static Queue<PulseData> PulseQueue = new Queue<PulseData>();
            public static int lowPulses = 0, highPulses = 0;
            public static Dictionary<string, Node> nodes;
            public string name;
            public string[] dests = new string[0];
            bool connectionsSet = false;
            public virtual void Pulse(bool highPulse, string source)
            {
            }

            public virtual void SetConnections(string input)
            {
                if (connectionsSet)
                    return;

                connectionsSet = true;

                foreach (string dest in dests)
                {
                    if (!nodes.ContainsKey(dest))
                    {
                        Node node = new Node();
                        node.name = input;
                        nodes[dest] = node;
                    }
                    nodes[dest].SetConnections(name);
                }
            }
        }

        class FlipFlop : Node
        {
            public bool on = false;

            public override void Pulse(bool highPulse, string source)
            {
                base.Pulse(highPulse, source);

                if (highPulse)
                    return;

                on = !on;

                foreach (string dest in dests)
                    SendPulse(on, name, dest);
            }
        }

        class Conjunction : Node
        {
            Dictionary<string, bool> inputs = new Dictionary<string, bool>();

            //public bool high = false;
            public override void Pulse(bool highPulse, string source)
            {
                base.Pulse(highPulse, source);

                inputs[source] = highPulse;

                bool high = true;
                foreach (string input in inputs.Keys)
                {
                    high &= inputs[input];
                }

                foreach (string dest in dests)
                    SendPulse(!high, name, dest);
            }

            public override void SetConnections(string input)
            {
                inputs[input] = false;
                base.SetConnections(input);
            }
        }

        class Broadcaster : Node
        {
            public override void Pulse(bool highPulse, string source)
            {
                base.Pulse(highPulse, source);
                foreach (string dest in dests)
                {
                    SendPulse(highPulse, name, dest);
                }
            }
        }

        class Button : Node
        {
            public override void Pulse(bool highPulse, string source)
            {
                base.Pulse(false, name);
            }
        }

        Dictionary<string, Node> nodes = new Dictionary<string, Node>();

        public void Part1()
        {
            string[] lines = File.ReadAllLines("input20.txt");

            foreach (string line in lines)
            {
                string name = line.Substring(0, line.IndexOf(' '));
                string destStr = line.Substring(line.IndexOf('>') + 1);
                string[] dests = destStr.Split(',', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);

                Node.nodes = nodes;

                Node node = null;

                if (name == "broadcaster")
                {
                    node = new Broadcaster();
                    node.name = name;
                }
                else if (name[0] == '%')
                {
                    node = new FlipFlop();
                    node.name = name.Substring(1);
                }
                else if (name[0] == '&')
                {
                    node = new Conjunction();
                    node.name = name.Substring(1);
                }
                else
                {
                    continue;
                }
                node.dests = dests;

                nodes[node.name] = node;
            }

            nodes["broadcaster"].SetConnections("button");

            for (int i = 0; i < 1000; i++)
            {
                //nodes["broadcaster"].Pulse(false, "button");
                Node.SendPulse(false, "button", "broadcaster");
                while (Node.ProcessQueue()) { }
            }

            Console.WriteLine($"High pulses: {Node.highPulses} Low: {Node.lowPulses}");
            int product = Node.highPulses * Node.lowPulses;
            Console.WriteLine($"Product: {product}");

            // 675553410 too low
        }

        public void Part2()
        {
            string[] lines = File.ReadAllLines("testinput20.txt");
        }
    }
}

File: AdventOfCode2023CPP/Day20.cpp
#include "Day20.h"

#include <string>
#include <queue>
#include <map>
#include <vector>
#include <fstream>
#include <iostream>

namespace {
    struct PulseData
    {
        std::string source;
        std::string dest;
        bool high;

        PulseData(const std::string &_source, const std::string &_dest, bool _high)
            : source(_source), dest(_dest), high(_high)
        {
        }
    };

    std::queue<PulseData> pulseQueue;
    int highPulses = 0, lowPulses = 0;

    class Node;
    class Conjunction;

    // Part 2
    bool monitoring = false;
    Conjunction *junction = nullptr;
    std::map<std::string, bool> monitorFound;
    std::map<std::string, int> monitorCounts;
    int buttonCount = 0;


    static std::map<std::string, Node *> nodes;

    class Node
    {
    public:
        std::string name;
        std::vector<std::string> dests;

        Node(const std::string &_name)
            : name(_name)
        {
        }

        static void SendPulse(bool high, std::string source, std::string dest)
        {
            PulseData data(source, dest, high);
            pulseQueue.push(data);

            if (nodes.find(dest) == nodes.end())
            {
                std::cout << "error" << std::endl;
            }

            if (high)
                highPulses++;
            else
                lowPulses++;
        }

        virtual void Pulse(bool high, const std::string &source)
        {
        }

        virtual void SetConnections(const std::string &input)
        {
            if (connectionsSet)
                return;

            connectionsSet = true;

            for (auto &dest : dests) {
                if (nodes.count(dest) == 0) {
                    Node *node = new Node(input);
                    nodes[dest] = node;
                }
                nodes[dest]->SetConnections(name);
            }
        }

    protected:
        bool connectionsSet = false;
    };

    class FlipFlop : public Node
    {
    public:
        bool on = false;

        FlipFlop(const std::string &_name)
            : Node(_name) {}

        void Pulse(bool highPulse, const std::string &source)
        {
            if (highPulse)
                return;

            on = !on;

            for (auto &dest : dests)
                SendPulse(on, name, dest);
        }
    };

    class Conjunction : public Node
    {
    public:
        std::map<std::string, bool> inputs;

        Conjunction(const std::string &_name)
            : Node(_name) {}

        void Pulse(bool highPulse, const std::string &source)
        {
            if (inputs.find(source) == inputs.end())
            {
                std::cout << "error" << std::endl;
            }
            //inputs[source] = !inputs[source];
            inputs[source] = highPulse;

            bool high = true;
            for (auto &input : inputs)
                high &= input.second;

            for (auto &dest : dests)
                SendPulse(!high, name, dest);
        }

        void SetConnections(const std::string &input)
        {
            inputs[input] = false;
            Node::SetConnections(input);
        }
    };

    class Broadcaster : public Node
    {
    public:
        Broadcaster(const std::string &_name)
            : Node(_name) {}

        void Pulse(bool highPulse, const std::string &source)
        {
            for (auto &dest : dests)
            {
                SendPulse(highPulse, name, dest);
            }
        }
    };

    bool ProcessQueue()
    {
        if (pulseQueue.empty())
            return false;

        PulseData &data = pulseQueue.front();

        nodes[data.dest]->Pulse(data.high, data.source);
        pulseQueue.pop();

        return true;
    }

}

void Day20::Setup(const char *inputfile)
{
    std::ifstream file(inputfile, std::ifstream::in);
    std::string str;

    while (std::getline(file, str)) {
        std::string name = str.substr(0, str.find(' '));
        std::string destStr = str.substr(str.find('>') + 2);
        std::vector<std::string> dests;

        size_t i = 0;
        size_t ii;
        do {
            ii = destStr.find(',', i);
            dests.push_back(destStr.substr(i, ii - i));
            i = ii + 2;
        } while (i < destStr.size() && ii != destStr.npos);

        Node *node = nullptr;

        if (name == "broadcaster")
        {
            node = new Broadcaster(name);
        }
        else if (name[0] == '%')
        {
            node = new FlipFlop(name.substr(1));
        }
        else if (name[0] == '&')
        {
            node = new Conjunction(name.substr(1));
        }
        else
        {
            continue;
        }
        node->dests = dests;

        nodes[node->name] = node;

        //std::cout << "Data: " << entry->data << " Pattern: " << pattern << std::endl;
    }

    nodes["broadcaster"]->SetConnections("button");
}

void Day20::Part1()
{
    Setup("input20.txt");

    for (int i = 0; i < 1000; i++) {
        Node::SendPulse(false, "button", "broadcaster");

        while (ProcessQueue()) {
        }
    }

    // too low - Low: 15030 High: 44947 Product: 675553410

    std::cout << "Low: " << lowPulses << " High: " << highPulses << " Product: " << (lowPulses * highPulses) << std::endl;
}

void Day20::Part2()
{
    Setup("input20.txt");

    for (auto &i : nodes) {
        if (!i.second->dests.empty() && i.second->dests[0] == "rx") {
             junction = static_cast<Conjunction *>(i.second);
             break;
        }
    }

    if (junction == nullptr)
        return;

    for (auto &i : junction->inputs) {
        monitorFound[i.first] = false;
        monitorCounts[i.first] = 0;
    }

    bool done = true;
    buttonCount = 0;
    do {
        Node::SendPulse(false, "button", "broadcaster");
        buttonCount++;

        while (ProcessQueue()) {
            for (auto &input : junction->inputs) {
                if (monitorFound[input.first])
                    continue;

                if (input.second) {
                    monitorFound[input.first] = true;
                    monitorCounts[input.first] = buttonCount;
                }
            }
        }

        done = true;
        for (auto f : monitorFound)
            done &= f.second;
    } while (!done);

    long long product = 1;
    for (auto &count : monitorCounts) {
        std::cout << count.first << " repeats after " << count.second << std::endl;
        product *= count.second;
        //std::cout << "Product = " << product << std::endl;
    }

    std::cout << "LCM = " << product << std::endl;
}
